lib/lrama/state/reduce.rb

```diff
@@ -26,8 +26,7 @@ def add_not_selected_symbol(sym)
       end
 
       def selected_look_ahead
-        if @look_ahead
-          # @type ivar @look_ahead: Array<Grammar::Symbol>
```

Comment:
I suspect `lib/lrama/state/reduce.rb:30:36: [error] Type annotation has a syntax error: Failed to parse a type in annotation`  might be a bug on steep. Do you have any idea?

Reply:
For example, these diff makes the code syntax valid, even so type check is failed.

```diff
diff --git a/lib/lrama/state/reduce.rb b/lib/lrama/state/reduce.rb
index a2b7c26..0e67b77 100644
--- a/lib/lrama/state/reduce.rb
+++ b/lib/lrama/state/reduce.rb
@@ -27,7 +27,7 @@ module Lrama

       def selected_look_ahead
         if @look_ahead
-          # @type ivar @look_ahead: Array<Grammar::Symbol>
+          # @type ivar @look_ahead: Array
           @look_ahead - @not_selected_symbols
         else
           []
```

Reply:
I applied monkey patches like below to `steep-1.9.1/lib/steep/annotation_parser.rb`.
It seems RBS parses codes without syntax error however `type` and `string` have different value.
So that "Failed to parse a type in annotation" SyntaxError is raised.

```ruby
    def parse_type(match, name = :type, location:)
      string = match[name] or raise
      st, en = match.offset(name)
      st or raise
      en or raise
      loc = RBS::Location.new(location.buffer, location.start_pos + st, location.start_pos + en)

      type =
        begin
          RBS::Parser.parse_type(string)
        rescue RBS::ParsingError => exn
          p exn
          raise SyntaxError.new(source: string, location: loc, exn: exn)
        end or raise

      unless (type.location || raise).source == string.strip
        p type
        p type.location
        p type.location.source
        p string
        raise SyntaxError.new(source: string, location: loc, message: "Failed to parse a type in annotation")
      end

      factory.type(type)
    end
```

```shell
$ bundle exec steep check
# Type checking files:
...
#<RBS::Types::ClassInstance:0x000000012703b9e8 @name=#<RBS::TypeName:0x000000012703ba38 @namespace=#<RBS::Namespace:0x000000012703ba60 @path=[], @absolute=false>, @name=:Array, @kind=:class>, @args=[], @location=#<RBS::Location:32740 buffer=a.rbs, start=1:0, pos=0...5, children=name,?args source="Array">>
#<RBS::Location:32740 buffer=a.rbs, start=1:0, pos=0...5, children=name,?args source="Array">
"Array"
"Array<Grammar::Symbol>"
```

Reply:
I notice that type parameter should be wrapped with `[]` not `<>`, e.g. `Array[Grammar::Symbol]`.

Reply:
I tried changing it to use `[]`, but it doesn't seem to be working as intended.
It looks the same as the behavior without any inline comments specified.

case1:
```diff
-# @type ivar @look_ahead: Array<Grammar::Symbol>
+# @type ivar @look_ahead: Array[Grammar::Symbol]
@look_ahead - @not_selected_symbols
```

result:
```
❯ bundle exec steep check
# Type checking files:

...............................................................................F.............

lib/lrama/state/reduce.rb:31:22: [error] Type `(::Array[::Lrama::Grammar::Symbol] | nil)` does not have method `-`
│ Diagnostic ID: Ruby::NoMethod
│
└           @look_ahead - @not_selected_symbols
                        ~

Detected 1 problem from 1 file
```
---

case2:
```diff
-# @type ivar @look_ahead: Array<Grammar::Symbol>
@look_ahead - @not_selected_symbols
```

result:
```
❯ bundle exec steep check
# Type checking files:

...............................................................................F.............

lib/lrama/state/reduce.rb:30:22: [error] Type `(::Array[::Lrama::Grammar::Symbol] | nil)` does not have method `-`
│ Diagnostic ID: Ruby::NoMethod
│
└           @look_ahead - @not_selected_symbols
                        ~

Detected 1 problem from 1 file
```

Reply:
It seems narrowing for instance variable might not work however for attr_reader works. 

```diff
diff --git a/Gemfile b/Gemfile
index 58c28f7..870fde0 100644
--- a/Gemfile
+++ b/Gemfile
@@ -15,6 +15,6 @@ gem "memory_profiler"
 # Recent steep requires Ruby >= 3.0.0.
 # Then skip install on some CI jobs.
 if !ENV['GITHUB_ACTION'] || ENV['INSTALL_STEEP'] == 'true'
-  gem "rbs", "3.6.0", require: false
-  gem "steep", "1.8.0", require: false
+  gem "rbs", "3.7.0", require: false
+  gem "steep", "1.9.1", require: false
 end
diff --git a/lib/lrama/state/reduce.rb b/lib/lrama/state/reduce.rb
index a2b7c26..54ab87b 100644
--- a/lib/lrama/state/reduce.rb
+++ b/lib/lrama/state/reduce.rb
@@ -26,9 +26,8 @@ module Lrama
       end

       def selected_look_ahead
-        if @look_ahead
-          # @type ivar @look_ahead: Array<Grammar::Symbol>
-          @look_ahead - @not_selected_symbols
+        if look_ahead
+          look_ahead - @not_selected_symbols
         else
           []
         end
```

Reply:
Indeed it is. Thank you very much. I have updated this PR.